mpris.c

```diff
@@ -0,0 +1,558 @@
+	if (strcmp(t, "None") == 0) {
+		player_repeat_current = 0;
+		repeat = 0;
+	} else if (strcmp(t, "Track") == 0) {
+		player_repeat_current = 1;
+	} else if (strcmp(t, "Playlist") == 0) {
+		player_repeat_current = 0;
+		repeat = 1;
+	}
+	update_statusline();
+	return sd_bus_reply_method_return(value, "");
+}
+
+static int mpris_rate(sd_bus *_bus, const char *_path,
+		const char *_interface, const char *_property,
+		sd_bus_message *reply, void *_userdata,
+		sd_bus_error *_ret_error)
+{
+	static const double d = 1.0;
+	return sd_bus_message_append_basic(reply, 'd', &d);
+}
+
+static int mpris_shuffle(sd_bus *_bus, const char *_path,
+		const char *_interface, const char *_property,
+		sd_bus_message *reply, void *_userdata,
+		sd_bus_error *_ret_error)
+{
+	const uint32_t s = shuffle;
+	return sd_bus_message_append_basic(reply, 'b', &s);
+}
+
+static int mpris_set_shuffle(sd_bus *_bus, const char *_path,
+		const char *_interface, const char *_property,
+		sd_bus_message *value, void *_userdata,
+		sd_bus_error *_ret_error)
+{
+	uint32_t s = 0;
+	CK(sd_bus_message_read_basic(value, 'b', &s));
+	shuffle = s;
+	update_statusline();
+	return sd_bus_reply_method_return(value, "");
+}
+
+static int mpris_volume(sd_bus *_bus, const char *_path,
+		const char *_interface, const char *_property,
+		sd_bus_message *reply, void *_userdata,
+		sd_bus_error *_ret_error)
+{
+	double vol;
+	if (soft_vol) {
+		vol = (soft_vol_l + soft_vol_r) / 200.0;
+	} else if (volume_max && volume_l >= 0 && volume_r >= 0) {
+		int vol_left = scale_to_percentage(volume_l, volume_max);
+		int vol_right = scale_to_percentage(volume_r, volume_max);
+		vol = (vol_left + vol_right) / 200.0;
+	}
+	return sd_bus_message_append_basic(reply, 'd', &vol);
+}
+
+static int mpris_set_volume(sd_bus *_bus, const char *_path,
+		const char *_interface, const char *_property,
+		sd_bus_message *value, void *_userdata,
+		sd_bus_error *_ret_error)
+{
+	double vol;
+	CK(sd_bus_message_read_basic(value, 'd', &vol));
+	if (vol < 0.0)
+		vol = 0.0;
+	else if (vol > 1.0)
+		vol = 1.0;
+	int ivol = vol * 100;
+	player_set_vol(ivol, VF_PERCENTAGE, ivol, VF_PERCENTAGE);
+	update_statusline();
+	return sd_bus_reply_method_return(value, "");
+}
+
+static int mpris_position(sd_bus *_bus, const char *_path,
+		const char *_interface, const char *_property,
+		sd_bus_message *reply, void *_userdata,
+		sd_bus_error *_ret_error)
+{
+	player_info_lock();
+	int64_t pos = player_info.pos;
+	player_info_unlock();
+	pos *= 1000 * 1000;
+	return sd_bus_message_append_basic(reply, 'x', &pos);
+}
+
+static int mpris_msg_append_simple_dict(sd_bus_message *m, const char *tag,
+		char type, const void *val)
+{
+	const char s[] = { type, 0 };
+	CK(sd_bus_message_open_container(m, 'e', "sv"));
+	CK(sd_bus_message_append_basic(m, 's', tag));
+	CK(sd_bus_message_open_container(m, 'v', s));
+	CK(sd_bus_message_append_basic(m, type, val));
+	CK(sd_bus_message_close_container(m));
+	CK(sd_bus_message_close_container(m));
```

Comment:
It's unnecessary. The library will discard the reply if we return a negative error code and free all contents.